File: app/Services/ImportanceScoreService.php
<?php

namespace App\Services;

use App\GithubConfig;
use App\Models\Item;
use Carbon\Carbon;

class ImportanceScoreService
{
    /**
     * Calculate the importance score for an item
     * Score is based on rules defined in GithubConfig::IMPORTANCE_SCORING
     */
    public static function calculateScore(Item $item): int
    {
        $score = 0;

        // Hard filter: item must be assigned to current user
        if (!$item->isCurrentlyAssignedToUser()) {
            return -999; // Return very low score for non-assigned items
        }

        // Check Project Board Status
        $score += self::getProjectStatusPoints($item);

        // Check Hotfix Friday
        $score += self::getHotfixFridayPoints($item);

        // Check Milestone Proximity
        $score += self::getMilestoneProximityPoints($item);

        // Check Review Status
        $score += self::getReviewStatusPoints($item);

        // Default points for items without milestone
        if (!$item->milestone_id) {
            $score += GithubConfig::IMPORTANCE_SCORING['without_milestone']['default_points'];
        }

        return $score;
    }

    /**
     * Get points based on project board status (In Progress/In Review)
     */
    private static function getProjectStatusPoints(Item $item): int
    {
        $config = GithubConfig::IMPORTANCE_SCORING['project_board_status'];

        if (!$config['enabled']) {
            return 0;
        }

        try {
            $projectItem = \DB::table('project_items')
                ->where('item_id', $item->id)
                ->first();

            if (!$projectItem) {
                return 0;
            }

            $status = \DB::table('project_item_field_values')
                ->where('project_item_id', $projectItem->id)
                ->where('field_name', 'Status')
                ->value('field_value');

            if (!$status) {
                return 0;
            }

            $isInProgress = collect($config['in_progress_keywords'])->some(fn($keyword) => str_contains(strtolower($status), $keyword));

            return $isInProgress ? $config['in_progress_points'] : 0;
        } catch (\Exception $e) {
            // Tables don't exist, return 0
            return 0;
        }
    }

    /**
     * Get points based on Hotfix Friday rules
     */
    private static function getHotfixFridayPoints(Item $item): int
    {
        $config = GithubConfig::IMPORTANCE_SCORING['hotfix_friday'];

        if (!$config['enabled']) {
            return 0;
        }

        $today = Carbon::now()->dayOfWeek; // 0=Sunday, 5=Friday

        if ($today !== $config['day']) {
            return 0;
        }

        // Check if item has the hotfix label
        if (self::hasLabel($item, $config['label'])) {
            return $config['points_when_active'];
        }

        return 0;
    }

    /**
     * Get points based on how close the milestone due date is
     */
    private static function getMilestoneProximityPoints(Item $item): int
    {
        $config = GithubConfig::IMPORTANCE_SCORING['milestone_proximity'];

        if (!$config['enabled'] || !$item->milestone) {
            return 0;
        }

        $dueDate = Carbon::parse($item->milestone->due_on);
        $now = Carbon::now();
        $daysUntilDue = $now->diffInDays($dueDate, false); // Can be negative

        // Find the closest matching days threshold
        $points = 0;
        foreach ($config['points_by_days_until_due'] as $days => $pointsValue) {
            if ($daysUntilDue <= $days && $daysUntilDue > 0) {
                $points = $pointsValue;
                break;
            }
        }

        return $points;
    }

    /**
     * Get points based on review status
     */
    private static function getReviewStatusPoints(Item $item): int
    {
        $config = GithubConfig::IMPORTANCE_SCORING['review_status'];

        if (!$config['enabled'] || !$item->isPullRequest()) {
            return 0;
        }

        $reviewStatus = self::getPullRequestReviewStatus($item);

        return match ($reviewStatus) {
            'pending' => $config['pending_review_points'],
            'changes_requested' => $config['changes_requested_points'],
            'approved' => $config['all_approved_points'],
            default => 0,
        };
    }

    /**
     * Determine the review status of a pull request
     */
    private static function getPullRequestReviewStatus(Item $item): string
    {
        if (!$item->isPullRequest()) {
            return 'none';
        }

        // Get reviews from base_comments table where type='review'
        $reviews = \App\Models\BaseComment::where('issue_id', $item->id)
            ->where('type', 'review')
            ->get()
            ->mapWithKeys(function($review) {
                return [$review->id => $review->reviewDetails];
            })
            ->filter();

        if ($reviews->isEmpty()) {
            return 'pending';
        }

        // Check if any review has changes requested
        $hasChangesRequested = $reviews->contains(fn($review) => $review->state === 'CHANGES_REQUESTED');
        if ($hasChangesRequested) {
            return 'changes_requested';
        }

        // Check if any review has comments
        $hasComments = $reviews->contains(fn($review) => $review->state === 'COMMENTED');
        if ($hasComments) {
            return 'changes_requested'; // Treat comments as actionable
        }

        // Check if all reviews are approved
        $allApproved = $reviews->every(fn($review) => $review->state === 'APPROVED');
        if ($allApproved) {
            return 'approved';
        }

        return 'pending';
    }

    /**
     * Check if an item has a specific label
     */
    private static function hasLabel(Item $item, string $labelName): bool
    {
        if (!is_array($item->labels)) {
            return false;
        }

        return in_array($labelName, $item->labels);
    }

    /**
     * Update the importance score for an item and save it
     */
    public static function updateItemScore(Item $item): void
    {
        $item->importance_score = self::calculateScore($item);
        $item->save();
    }

    /**
     * Recalculate and update scores for multiple items
     */
    public static function updateMultipleItemScores(array $items): void
    {
        foreach ($items as $item) {
            self::updateItemScore($item);
        }
    }
}

File: app/services/importance_score_service.rb
# Recomputes items.importance_score — the shared "stats" column the Laravel
# dashboard sorts by (`where importance_score > 0 order by ... desc`). Ported
# from Laravel's ImportanceScoreService; every webhook job that changes a
# scoring input calls update_item_score, mirroring the Laravel listeners.
#
# One deliberate divergence: review status is derived from the
# RequestedReviewer state machine (per-reviewer latest verdict, dismissals
# revert to pending) instead of Laravel's scan over all historical review
# rows — Laravel compared states in the wrong case, so its review category
# silently scored every PR as "pending". This is what makes dismissing or
# re-requesting a review actually move the score.
class ImportanceScoreService
  EXCLUDED_SCORE = -999

  class << self
    def update_item_score(item)
      item.update_column(:importance_score, calculate_score(item))
    end

    def calculate_score(item)
      config = GithubConfig::IMPORTANCE_SCORING

      # Hard filters: only items assigned to the configured user, without an
      # excluded label, compete for a score.
      return EXCLUDED_SCORE unless item.assigned_to_configured_user?
      return EXCLUDED_SCORE if (label_names(item) & config[:filters][:excluded_labels]).any?

      config[:category_weights].sum { |category, weight|
        send("#{category}_score", item) * weight / 100.0
      }.round
    end

    # --- Categories (each returns 0..100, except review's -20 floor) --------

    def milestone_urgency_score(item)
      config = GithubConfig::IMPORTANCE_SCORING
      due_on = item.milestone&.due_on
      return config[:without_milestone][:normalized_score] if due_on.nil?

      days_until_due = (due_on.to_date - Date.current).to_i

      if days_until_due.negative?
        overdue = config[:milestone_proximity][:overdue]
        escalation = [ days_until_due.abs * overdue[:escalation_per_day], 100 - overdue[:normalized_score] ].min
        return [ overdue[:normalized_score] + escalation, 100 ].min
      end

      range = config[:milestone_proximity][:ranges].find { |r| days_until_due.between?(r[:min_days], r[:max_days]) }
      range ? range[:normalized_score] : 0
    end

    def review_status_score(item)
      config = GithubConfig::IMPORTANCE_SCORING[:review_status]
      return 0 unless item.pull_request?

      case pull_request_review_status(item)
      when "changes_requested" then config[:changes_requested_normalized]
      when "approved" then config[:approved_normalized]
      else config[:pending_review_normalized]
      end
    end

    def unresolved_comments_score(item)
      config = GithubConfig::IMPORTANCE_SCORING[:unresolved_comments]

      comments = BaseComment.unscoped
        .where(issue_id: item.id, type: "code", resolved: false)
        .includes(:github_user)
      return 0 if comments.empty?

      count = comments.sum do |comment|
        comment.github_user&.name == config[:critical_reviewer] ? config[:critical_count_multiplier] : 1
      end

      [ count.to_f / config[:max_score_at_count] * 100, 100 ].min.round
    end

    def project_board_status_score(item)
      config = GithubConfig::IMPORTANCE_SCORING[:project_board_status]

      status = project_status(item)
      return 0 if status.blank?

      in_progress = config[:in_progress_keywords].any? { |keyword| status.downcase.include?(keyword.downcase) }
      in_progress ? config[:normalized_score] : 0
    end

    def hotfix_friday_score(item)
      config = GithubConfig::IMPORTANCE_SCORING[:hotfix_friday]
      return 0 unless Date.current.wday == config[:day]

      label_names(item).include?(config[:label]) ? config[:normalized_score] : 0
    end

    # "changes_requested" while any reviewer's latest verdict blocks;
    # "approved" once someone approved and nobody is still pending;
    # "pending" otherwise (no reviews yet, or only comments so far).
    def pull_request_review_status(item)
      states = RequestedReviewer.where(pull_request_id: item.id).pluck(:state)
      return "changes_requested" if states.include?("changes_requested")
      return "approved" if states.include?("approved") && !states.include?("pending")

      "pending"
    end

    private

    def label_names(item)
      item.labels.pluck(:name)
    end

    PROJECT_STATUS_QUERY = <<~GRAPHQL
      query ($org: String!, $repo: String!, $number: Int!) {
        repository(owner: $org, name: $repo) {
          issueOrPullRequest(number: $number) {
            ... on Issue {
              projectItems(first: 10) {
                nodes { fieldValueByName(name: "Status") { ... on ProjectV2ItemFieldSingleSelectValue { name } } }
              }
            }
            ... on PullRequest {
              projectItems(first: 10) {
                nodes { fieldValueByName(name: "Status") { ... on ProjectV2ItemFieldSingleSelectValue { name } } }
              }
            }
          }
        }
      }
    GRAPHQL

    def project_status(item)
      repository = item.repository
      return nil if repository&.organization.nil?

      response = GithubApi.graphql(PROJECT_STATUS_QUERY,
        org: repository.organization.name,
        repo: repository.name,
        number: item.number)

      nodes = response.dig("data", "repository", "issueOrPullRequest", "projectItems", "nodes") || []
      nodes.first&.dig("fieldValueByName", "name")
    rescue GithubApi::Error
      nil
    end
  end
end
